### PR TITLE
HFS/586 - Get transactions

### PR DESCRIPTION
**Pull request by gertaylor**
## Link to JIRA ticket

Add a link to the JIRA ticket that the changes in this PR describe.

### *What is the problem we're trying to solve*

Add a summary of the problem and any additional context that helps explain what the issue we're trying to solve is.

### *What changes have we introduced*

Add descriptions of what changes were made to address the problem and provide a solution.

Include any links to commits that introduce significant additions of code if they help explain the process of coming to the solution. e.g Addition of test database setup, addition of first test and production class, removal of buggy code, etc.

#### _Checklist_

- [ ] Added end-to-end (i.e. integration) tests where necessary e.g to test complete functionality of a newly added feature
- [ ] Added tests to cover all new production code
- [ ] Added comments to the README or updated relevant documentation _(add link to documentation)_, where necessary.
- [ ] Checked all code for possible refactoring
- [ ] Code pipeline builds correctly

### *Follow up actions after merging PR*

Are there any next steps that need to addressed after merging this PR? Add them here.

### CODE REVIEW

File: FinancialTransactionsApi/V1/UseCase/GetAllUseCase.cs [hunk 1]
```diff
@@ -17,13 +18,11 @@ public GetAllUseCase(ITransactionGateway gateway)
             _gateway = gateway;
         }
 
-        public async Task<PagedResult<TransactionResponse>> ExecuteAsync(TransactionQuery query)
+        public async Task<PaginatedResponse<TransactionResponse>> ExecuteAsync(TransactionQuery query)
         {
-            var transactions =
-                await _gateway.GetPagedTransactionsAsync(query)
-                    .ConfigureAwait(false);
+            Paginated<Transaction> result = await _gateway.GetPagedTransactionsAsync(query).ConfigureAwait(false);
```
Comment (FaisalHackney) [inline FinancialTransactionsApi/V1/UseCase/GetAllUseCase.cs]: maybe change this to 
`var transactions = await _gateway.GetPagedTransactionsAsync(query).ConfigureAwait(false);`

File: FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs [hunk 2]
```diff
@@ -38,7 +38,31 @@ public async Task<Transaction> GetTransactionByIdAsync(Guid id)
             return await Task.FromResult(response.FirstOrDefault()?.ToDomain()).ConfigureAwait(false);
         }
 
-        public Task<PagedResult<Transaction>> GetPagedTransactionsAsync(TransactionQuery query) => throw new NotImplementedException();
+        public async Task<Paginated<Transaction>> GetPagedTransactionsAsync(TransactionQuery query)
```
Comment (FaisalHackney) [inline FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs]: Does the  `<Paginated<Transaction>>`
implicitly return an `<IEnumerable<Transaction>`?

File: FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs [hunk 3]
```diff
@@ -38,7 +38,31 @@ public async Task<Transaction> GetTransactionByIdAsync(Guid id)
...
+                response = response.Where(x => x.TransactionType == query.TransactionType.ToString());
+            }
+
+            var page = query.Page > 1 ? query.Page : 0;
+
+            var filteredResponse = response.Skip(page * query.PageSize).Take(query.PageSize).Select(x => x.ToDomain());
+
+            var result = await Task.FromResult(filteredResponse.AsEnumerable()).ConfigureAwait(false);
+
+            return new Paginated<Transaction>()
```
Comment (FaisalHackney) [inline FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs]: Ideally the responsibility of the Gateway class is to call the database logic for querying and return the results to the UseCase class.
The responsibility of the UseCase class would be to do all the mapping to Paginated Results and returning that to the Controller.

**Review by FaisalHackney — CHANGES_REQUESTED** [review]

File: FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs (hunk 2 above)
Comment (gertaylor) [reply FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs]: Yes the results on the paginated object returns a ` IEnumerable<T> Results`

File: FinancialTransactionsApi/V1/UseCase/GetAllUseCase.cs (hunk 1 above)
Comment (gertaylor) [reply FinancialTransactionsApi/V1/UseCase/GetAllUseCase.cs]: updated

File: FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs (hunk 3 above)
Comment (gertaylor) [reply FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs]: I followed the logic of the GetAllActive method. Would you suggest the method just returns the `IEnumerable<Transactions>` and then the other logic is done in the UserCase?

Comment (FaisalHackney) [reply FinancialTransactionsApi/V1/Gateways/PostgreDbGateway.cs]: yes please

**Review by FaisalHackney — APPROVED** [review]

**Pull request merged**